kernel/byterun/coq_interp.c

```diff
@@ -230,6 +230,12 @@ if (sp - num_args < coq_stack_threshold) {                                     \
     *sp = swap_accu_sp_tmp__;                   \
   }while(0)
 
+/* Turn a code pointer into a stack value usable as a return address, and conversely.
+   The least significant bit is set to 1 so that the GC does not mistake return
+   addresses for heap pointers. */
+#define StoreRA(p) ((value)p + 1)
```

Comment:
Is this legal C semantics? If I read the OCaml code correctly, `value := intnat` and `code_t := int32_t*`. C lets you create pointers that point one value past the valid memory, but I am not sure about the validity of the casts performed here.

Reply:
This is legal. C forbids you to create nonsensical pointers, but here, the arithmetic happens on integers, not on pointers. Indeed, `(value)p + 1` means that the pointer is first converted to an integer, which OCaml promises us is large enough to hold a pointer, and then this integer is incremented by one.